### PR TITLE
feat: rate limiter

### PR DESCRIPTION
**Pull request by montekki**
idk

### CODE REVIEW

File: core/sonm.go [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+const (
+	dwhUpdateInterval = 5 * time.Second
+)
+
+var accountSlotsDefaults = map[uint64]uint64{
+	0: 16,
+	1: 32,
+	2: 64,
+	3: 128,
+	4: 256,
```
Comment (nikonov1101) [inline core/sonm.go:29]: `0..4` are constants with identity levels? Please use protobuf's consts.

File: cmd/utils/flags.go [hunk 2]
```diff
@@ -289,6 +289,22 @@ var (
 		Usage: "Maximum amount of time non-executable transaction are queued",
 		Value: eth.DefaultConfig.TxPool.Lifetime,
 	}
+	// Sonm extensions settings
+	SonmDWHEndpointFlag = cli.StringFlag{
+		Name:  "sonm.dwhendpoint",
+		Usage: "Address of the SONM DWH geth sould use",
```
Comment (nikonov1101) [inline cmd/utils/flags.go:295]: mistype in `should`?

File: core/sonm.go [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+)
+
+const (
+	dwhUpdateInterval = 5 * time.Second
+)
+
+var accountSlotsDefaults = map[uint64]uint64{
+	0: 16,
```
Comment (nikonov1101) [inline core/sonm.go:25]: AFAIR zero is an invalid value for identity level, and 1 is for anonymous.

File: core/sonm.go [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	log.Info("Updaring SONM account level infos")
+
+	for acc, level := range m.accLevels {
+		profile, err := m.dwh.GetProfileInfo(m.ctx, &sonm.EthID{Id: sonm.NewEthAddress(acc)})
+
+		if err != nil {
+			continue
```
Comment (nikonov1101) [inline core/sonm.go:106]: I'd prefer also log an error (just to be sure that DWH is not fucked)

File: core/sonm.go [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+}
+
+func (m *sonmExtension) AccountSlots(addr common.Address) uint64 {
+	m.mu.RLock()
+	defer m.mu.Unlock()
+
+	level, ok := m.accLevels[addr]
+	if !ok {
+		sonmAddr := sonm.NewEthAddress(addr)
+		profile, err := m.dwh.GetProfileInfo(m.ctx, &sonm.EthID{Id: sonmAddr})
```
Comment (nikonov1101) [inline core/sonm.go:78]: use context with timeout to limit the request time

File: core/sonm.go [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+	KeyPath     string
+	KeyPass     string
+	DWHEndpoint string
+}
+
+var DefaultSonmConfig = SonmConfig{
+	DWHEndpoint: "0xadffcac607a0a1b583c489977eae413a62d4bc73@dwh.livenet.sonm.com:15021",
+}
+
+type SonmExtension interface {
```
Comment (nikonov1101) [inline core/sonm.go:42]: what the reason to introduce interface?

File: core/sonm.go [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+
+	dwhCC, err := xgrpc.NewClient(se.ctx, cfg.DWHEndpoint, transportCredentials)
+	if err != nil {
+		return nil, err
+	}
+
+	se.dwh = sonm.NewDWHClient(dwhCC)
+
+	se.wg.Add(1)
+	go se.loop()
```
Comment (nikonov1101) [inline core/sonm.go:156]: consider removing waitgroup and make a start method that can be called somewhere after module construction.

File: core/sonm.go [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+	0: 16,
+	1: 32,
+	2: 64,
+	3: 128,
+	4: 256,
+}
+
+type SonmConfig struct {
+	KeyPath     string
+	KeyPass     string
```
Comment (nikonov1101) [inline core/sonm.go:34]: AFAIK, geth-node in mining mode already uses a keystore, can we use it instead of opening another one?

**Review by nikonov1101 — CHANGES_REQUESTED** [review]

**Comment by sokel** [comment]
please form header and description of the pr

File: core/sonm.go (hunk 8 above)
Comment (sokel) [reply core/sonm.go:34]: these option useful without mining mode.
but there, geth have code for workin with key. right.

File: core/sonm.go [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+	}
+
+	return accountSlotsDefaults[0]
+}
+
+func (m *sonmExtension) updateLevelInfo() {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	log.Info("Updaring SONM account level infos")
```
Comment (sokel) [inline core/sonm.go:100]: typo Upda**t**ing

**Comment by sokel** [comment]
nead to alter and repair

**Pull request closed without merging**